apps/shared/avifjpeg.c

```diff
@@ -354,24 +383,134 @@ avifBool avifJPEGRead(const char * inputFilename,
+                if ((uint64_t)standardXMP.size + totalExtendedXMPSize > SIZE_MAX) {
+                    fprintf(stderr, "XMP extraction failed: total XMP size is too high\n");
+                    goto cleanup;
+                }
+                if ((extendedXMPSize == 0) || ((uint64_t)extendedXMPOffset + extendedXMPSize > totalExtendedXMPSize)) {
+                    fprintf(stderr, "XMP extraction failed: invalid extended XMP segment size or offset\n");
+                    goto cleanup;
+                }
+                if (foundExtendedXMP) {
+                    if (memcmp(guid, extendedXMPGUID, extendedXMPGUIDLength)) {
+                        fprintf(stderr, "XMP extraction failed: extended XMP segment GUID mismatch\n");
+                        goto cleanup;
+                    }
+                    if (totalExtendedXMPSize != totalXMP.size) {
+                        fprintf(stderr, "XMP extraction failed: extended XMP total size mismatch\n");
+                        goto cleanup;
+                    }
+                } else {
+                    memcpy(extendedXMPGUID, guid, extendedXMPGUIDLength);
+
+                    avifRWDataRealloc(&totalXMP, standardXMP.size + (size_t)totalExtendedXMPSize);
+                    memcpy(totalXMP.data, standardXMP.data, standardXMP.size);
+
+                    // Keep track of the bytes that were set.
+                    avifRWDataRealloc(&extendedXMPReadBytes, (size_t)totalExtendedXMPSize);
+                    memset(extendedXMPReadBytes.data, AVIF_FALSE, extendedXMPReadBytes.size);
+
+                    foundExtendedXMP = AVIF_TRUE;
+                }
+                // According to Adobe XMP Specification Part 3 section 1.1.3.1:
+                //   "A robust JPEG reader should tolerate the marker segments in any order."
+                memcpy(&totalXMP.data[standardXMP.size + extendedXMPOffset], &marker->data[offsetTillExtendedXMP], extendedXMPSize);
+
+                // Make sure no previously read data was overwritten by the current segment.
+                if (memchr(&extendedXMPReadBytes.data[extendedXMPOffset], AVIF_TRUE, extendedXMPSize)) {
+                    fprintf(stderr, "XMP extraction failed: overlapping extended XMP segments\n");
+                    goto cleanup;
+                }
+                // Keep track of the bytes that were set.
+                memset(&extendedXMPReadBytes.data[extendedXMPOffset], AVIF_TRUE, extendedXMPSize);
+            }
+        }
+
+        if (foundExtendedXMP) {
+            // Make sure there is no missing byte.
+            if (memchr(extendedXMPReadBytes.data, AVIF_FALSE, extendedXMPReadBytes.size)) {
+                fprintf(stderr, "XMP extraction failed: missing extended XMP segments\n");
                 goto cleanup;
             }
+
+            // According to Adobe XMP Specification Part 3 section 1.1.3.1:
```

Comment:
Previously I had seen
https://archimedespalimpsest.net/Documents/External/XMP/XMPSpecificationPart3.pdf
where these are in chapter 3. What version are you using?

Adobe's links 404 currently:
  https://github.com/adobe/xmp-docs/blob/master/Specifications.md

Reply:
> What version are you using?

I do not remember where I got it but it seems to be the "January 2020" version as written in the front page.

> Previously I had seen https://archimedespalimpsest.net/Documents/External/XMP/XMPSpecificationPart3.pdf where these are in chapter 3.

Yes, the section "Embedding XMP metadata in application files" was moved from index 3. to index 1.

> Adobe's links 404 currently

Could this be because it is now an ISO specification behind a paywall? [This one](https://www.iso.org/standard/79384.html) seems related but maybe it is just an extension or something.

Reply:
It's possible the standardization could be the cause. Thanks for pointing me to your version.